compiler/rustc_codegen_llvm/src/abi.rs

```diff
@@ -546,6 +546,18 @@ impl<'tcx> FnAbiLlvmExt<'tcx> for FnAbi<'tcx, Ty<'tcx>> {
         if cconv != llvm::CCallConv {
             llvm::SetInstructionCallConv(callsite, cconv);
         }
+
+        if self.conv == Conv::CmseNonSecureCall {
+            // This will probably get ignored on all targets but those supporting the TrustZone-M
+            // extension (thumbv8m targets).
```

Comment:
We should reject the ABI when using a `--target` that doesn't support it (hmm, do we not do that for `#[cmse_non_secure_entry]`?)

Reply:
True! Initially I was looking how to check the target from this file but I am not sure how to get the target information from there. It looks like it would be possible to do it in the change file ` compiler/rustc_ast_passes/src/feature_gate.rs` as there is access to the `rustc_session::session::Session` structure which contains the target.

Reply:
If I do, can I maybe re-use and modify the same error code used for `cmse_nonsecure_entry`?

Reply:
Yeah that should be fine

Reply:
`feature_gate.rs` is not the right place to check this though. We already have per-target lists of supported ABIs, this should just plug in there.

Reply:
Oh, apparently we don't check ABIs at all, `avr-interrupt` compiles fine on x86, so scratch this for now

Reply:
Ok! Shall I still keep the `bare_function` thing there?

Reply:
Yeah we should avoid the LLVM assertion

Reply:
This should still be done here though: https://github.com/rust-lang/rust/blob/b05fd2a15de7c9e50110e9ed4c01f114be215739/compiler/rustc_typeck/src/check/check.rs#L34-L45

(not in the `feature_gate.rs` visitor, that's only for checking that the feature is enabled)

Reply:
It seems to check the `unsupported_abis` field of the `TargetOptions` struct of each target spec file. Shall I add `Abi::CCmseNonSecureCall` in each one of them but the `thumbv8m*` one? Just checking before I change everything 😋

Reply:
You only need to do the `bare_function` check there. The issue with unsupported ABIs being accepted is https://github.com/rust-lang/rust/issues/81396 and can be fixed later on.

Reply:
Ah I see, sorry I mixed up the two things and thought we were talking about the target check. Cool I can move the `bare_function` check there.

Just wondering how I would do it 😅 In the `feature_gate.rs` I can have the information that the compiler is currently checking a bare function type but I don't see how I can check that in the `check.rs` file. Or did you mean that I should call the `check_abi` function of `check.rs` (with the `bare_function` info) *from* the `feature_gate.rs` file?

Reply:
No, `feature_gate.rs` should not be involved in checking ABI-specific restrictions. I'm not sure where the best way to put this is, perhaps you have to write your own visitor, similar to what inline asm uses?

Reply:
Ah ok, got it.

I noticed that `check_abi` on `check.rs` is only called to check the ABI of function declarations and external blocks which are the exact two cases where the `CCmseNonSecureCall` ABI is forbidden. I believe the other only use of `extern` is for function pointer and that one is allowed.

So one easy way would be to add a check for `CCmseNonSecureCall` inside `check_abi` and emit an error if that's equal. I tested it and it works as expected.

Maybe it's a bit risky if `check_abi` is ever used for function pointers but we can add tests covering all the cases?

Reply:
Yeah, seems fine if there are tests

Reply:
Ok done, there should be:
* one test checking that the ABI on a function pointer compiles successfully
* one test checking that the ABI on on function definition fails
* one test checking that the ABI on an external block fails